plugins/notifications-backend/src/service/handlers.ts

```diff
@@ -0,0 +1,425 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* eslint-disable func-names */
+import { CatalogClient } from '@backstage/catalog-client';
+
+import { Knex } from 'knex';
+
+import { Components, Paths } from '../openapi';
+import {
+  DefaultMessageScope,
+  DefaultOrderBy,
+  DefaultOrderDirection,
+  DefaultPageNumber,
+  DefaultPageSize,
+} from './constants';
+import { ActionsInsert, dbValToBoolean, MessagesInsert } from './db';
+import {
+  MessageScopes,
+  NotificationsFilterRequest,
+  NotificationsOrderByDirections,
+  NotificationsOrderByFields,
+  NotificationsSortingRequest,
+} from './types';
+
+// createNotification
+// returns string id of created notification
+export async function createNotification(
+  dbClient: Knex<any, any>,
+  catalogClient: CatalogClient,
+  req: Paths.CreateNotification.RequestBody,
+): Promise<Paths.CreateNotification.Responses.$200> {
+  let isUser = false;
+
+  // validate users
+  if (Array.isArray(req.targetGroups) && req.targetGroups.length > 0) {
+    isUser = true;
+    const promises = req.targetGroups.map(group => {
+      return catalogClient.getEntityByRef(`group:${group}`).then(groupRef => {
```

Comment:
You could do `catalogClient.getEntitiesByRefs` to not call this for each entity

Reply:
done, thanks